src/gauges.rs

```diff
@@ -0,0 +1,316 @@
+
+        self.transaction_count
+            .set(epoch_info.transaction_count.unwrap_or_default() as i64);
+        self.slot_height.set(epoch_info.absolute_slot as i64);
+        self.current_epoch.set(epoch_info.epoch as i64);
+        self.current_epoch_first_slot.set(first_slot);
+        self.current_epoch_last_slot.set(last_slot);
+
+        Ok(())
+    }
+
+    // For now, this will export the IP addresses of active voting accounts with a node.
+    // TODO: This needs to actually export to a Prometheus gauge.
+    pub async fn export_ip_addresses(
+        &self,
+        nodes: &[RpcContactInfo],
+        vote_accounts: &RpcVoteAccountStatus,
+        api_key: &MaxMindAPIKey,
+        cache: &GeoCache,
+    ) -> anyhow::Result<()> {
+        // Define all types here
+        type RpcInfo = (RpcContactInfo, RpcVoteAccountInfo);
+        type RpcInfoMaybeGeo = (RpcContactInfo, RpcVoteAccountInfo, Option<CityApiResponse>);
+        type RpcInfoGeo = (RpcContactInfo, RpcVoteAccountInfo, CityApiResponse);
+
+        // All nodes that are validators
+        let validator_nodes = {
+            // Mapping of pubkey -> vote account info.
+            // Convert into a mapping to avoid O(n) searches
+            let pubkeys = vote_accounts
+                .current
+                .iter()
+                .cloned()
+                .map(|vote| (vote.node_pubkey.to_string(), vote))
+                .collect::<HashMap<String, RpcVoteAccountInfo>>();
+
+            nodes
+                .iter()
+                .cloned()
+                .filter_map(|contact| {
+                    pubkeys
+                        .get(&contact.pubkey)
+                        .map(|vote| (contact, vote.clone()))
+                })
+                .collect::<Vec<RpcInfo>>()
+        };
+
+        // Separate cached from uncached
+        let (cached, uncached): (Vec<RpcInfoMaybeGeo>, Vec<RpcInfoMaybeGeo>) = validator_nodes
+            .into_iter()
+            .map(|(contact, vote)| {
+                let cached = cache
+                    .fetch_ip_address_with_invalidation(
+                        &get_rpc_contact_ip(&contact).with_context(|| {
+                            format!("Validator node has no IP: {:?} {:?}", contact, vote)
+                        })?,
+                        |date| date + Duration::week() < OffsetDateTime::now_utc().date(),
+                    )?
+                    .map(|geo| geo.response);
+                Ok((contact, vote, cached))
+            })
+            .collect::<anyhow::Result<Vec<RpcInfoMaybeGeo>>>()?
+            .into_iter()
+            .partition(|(_, _, db)| db.is_some());
+
+        let mut geolocations = cached
+            .into_iter()
+            // This unwrap is safe because we have already partitioned into Some and Nones.
+            .map(|(c, v, db)| (c, v, db.unwrap()))
+            .collect::<Vec<RpcInfoGeo>>();
+
+        // For uncached, request them from maxmind.
+        println!(
```

Comment:
Could this be a `debug!` or an `info!`?

Reply:
I'm not too sure what's the difference between `debug!` and `info!` in terms of when they appear, or whether they will show up in a log-file? Speaking of, adding support for detailed logging would be very useful indeed (perhaps in a new PR)/

Reply:
debug would require `RUST_LOG=debug`. info would be printed by default without additional env.

Reply:
Detailed logging is already available with `RUST_LOG=debug`.

Reply:
I think in this case `debug` would be the appropriate level to log it at.

Reply:
Fixed in ab9d840